### ISSUE TITLE
Add EntityFramework Migrations Support

### ISSUE THREAD

**Issue by DavidNorena** [ExtCore/ExtCore#89]
Because of #72 and #88 and many others that are there and will come, I think is a good way to go and bring this feature to life together.

So after a research I found this https://github.com/aspnet/EntityFrameworkCore/issues/9033#issuecomment-313148393 for DbContext discovery:

> We always look in three places:
> 
> 1. Referenced by IDesignTimeDbContextFactory implementations in the assembly or startup assembly
> 2. Registered in the application's service provider
> 3. DbContext derived types in the assembly or startup assembly
>
> After we've discovered them all, we pick one to use.
> 
> I've considered having a way to bypass the type discovery (e.g. specify an assembly-qualified name on the command) and then discover a construction mechanism (factory first, service provider second, default constructor last), but this would require a completely separate (new) code path, and the value always seemed minimal.

From my point of view we can go for **IDesignTimeDbContextFactory Implementation** (because we have a "Virtual DbContext"), so we can override the construction of the DbContext and provide the IStorage.StorageContext that is derived from DbContext.

Something like this:

```
using ExtCore.Data.EntityFramework.PostgreSql;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;

namespace RadarJudicial.Data.App.Core
{
  public class DesignTimeDbContextFactory : IDesignTimeDbContextFactory<StorageContext>
  {
    public StorageContext CreateDbContext(string[] args)
    {
      // *** code for example purposes ***
      //var builder = new DbContextOptionsBuilder<StorageContext>();
      //builder.UseNpgsql(
          //"Host=localhost;Port=5432;User ID=radar;Password=radar123*;Database=radar;Pooling=true;");
      // *** code for example purposes ***
      
      // Here we need to Find resolve IStorage
      // Then find all Implementations of RepositoryBase
      // and/or seed every DbSet of the IStorage.StorageContext
      
      //Then return IStorage.StorageContext

      // Only for example purposes
      //return IStorage.StorageContext();
    }
  }
}
```

The bad news is that we cannot resolve services from here, and also this PoC is "attached" to PostgreSql, so we need a make a generic version of it .... what do you think @DmitrySikorsky does this give you more or less my idea ?

**Comment by DavidNorena** [comment]
https://github.com/aspnet/EntityFrameworkCore/issues/9467

**Comment by DmitrySikorsky** [comment]
Thank you for this great input! I will read it carefully and reply.

**Comment by DavidNorena** [comment]
Sure I want to be part of this good feature !

**Comment by DavidNorena** [comment]
Please checkout this: https://wildermuth.com/2017/07/06/Program-cs-in-ASP-NET-Core-2-0

**Comment by DmitrySikorsky** [comment]
I've created test project and playing with it right now. I will come back when have the results.

**Comment by DmitrySikorsky** [comment]
Thanks @DavidNorena, it seems that the IDesignTimeDbContextFactory interface is the best way to go. I can see that migrations tool runs the application (I can see the ExtCore logs) and than fails because it can't find the DbContext-derived class. If I add the -Project attribute with the ExtCore.Data.EntityFramework.Sqlite value (because there is the class inherited from the DbContext), it says that it can't instantiate it. Ok. So I created the implementation of the IDesignTimeDbContextFactory interface in the main web application.  What I'm thinking about is that EF tool runs the app, there is the real  storage context exists in the DI. We can access the DI from the IDesignTimeDbContextFactory as I understand for now, but I simple added the static property to my implementation of the IDesignTimeDbContextFactory interface and initialized it from the Configure Services like this:

`DesignTimeDbContextFactory.StorageContext = services.BuildServiceProvider().GetService<IStorageContext>() as StorageContextBase;`

In the implementation of the IDesignTimeDbContextFactory I simple return that value:

```
public class DesignTimeDbContextFactory : IDesignTimeDbContextFactory<StorageContextBase>
  {
    public static StorageContextBase StorageContext { get; set; }

    public StorageContextBase CreateDbContext(string[] args)
    {
      return DesignTimeDbContextFactory.StorageContext;

```
And it seems to partially work. I have next error while attempt to add an migration (but I don't have "dbcontext not found" error anymore:

```
Change your migrations assembly by using DbContextOptionsBuilder. E.g. options.UseSqlServer(connection, b => b.MigrationsAssembly("WebApplication")). By default, the migrations assembly is the assembly containing the DbContext.
Change your target project to the migrations project by using the Package Manager Console's Default project drop-down list, or by executing "dotnet ef" from the directory containing the migrations project.
Your target project 'WebApplication' doesn't match your migrations assembly 'ExtCore.Data.EntityFramework.Sqlite'. Either change your target project or change your migrations assembly.
Change your migrations assembly by using DbContextOptionsBuilder. E.g. options.UseSqlServer(connection, b => b.MigrationsAssembly("WebApplication")). By default, the migrations assembly is the assembly containing the DbContext.
Change your target project to the migrations project by using the Package Manager Console's Default project drop-down list, or by executing "dotnet ef" from the directory containing the migrations project.
```

I  think I should move the implementation of the IDesignTimeDbContextFactory to the package with the concrete storage implementations (like ExtCore.Data.EntityFramework.Sqlite etc) and set the package while adding migrations.

I think we could try to make this method work, because in this case we don't need to  pass connection strings and other parameters etc.

What do you think?

**Comment by DmitrySikorsky** [comment]
BTW. I can share my test project here.

**Comment by DavidNorena** [comment]
can you please share your test project ?

**Comment by DavidNorena** [comment]
> I think I should move the implementation of the IDesignTimeDbContextFactory to the package with the concrete storage implementations (like ExtCore.Data.EntityFramework.Sqlite etc) and set the package while adding migrations.

Yeah you are right, that way the user doesnt have to implement the class everytime, because the configuration will be made automagically by every provider the user uses ...

**Comment by DmitrySikorsky** [comment]
Here is my test app:

[ExtCore-Migrations.zip](https://github.com/ExtCore/ExtCore/files/1349227/ExtCore-Migrations.zip)

So our goal is to copy this IDesignTimeDbContextFactory implementation to 4 our storage providers and initialize the StorageContext property in the action, so user have to do nothing in order to make it all work.

Another question is should we do it as a solid part of the ExtCore? Maybe it should be optional? So somebody could write own implementation if need.

**Comment by DavidNorena** [comment]
Because its inside ExtCore.Data.EntityFramework and its obiously related to EntityFramework I think it should be a solid part of extcore, but what scenario do you have in mind, in which a user would need to write his/her own implementation ? what for ?

And Ok, Im gonna test your project, and will copy that implementation to all the storage providers then and send a PR as soon as I test everything !

**Comment by DmitrySikorsky** [comment]
I don't use migrations often, so its hard to say for me. We can do that and release version 3.1.0-beta1, so maybe somebody will tell us that it was wrong :)

**Comment by DavidNorena** [comment]
Ok then go for 3.1.0-beta-1. 

I think we should have only one implementation of DesignTimeDbContextFactory stored at ExtCore.Data.EntityFramework next to StorageContextBase it would be generic ... what do you think about it ?

**Comment by DmitrySikorsky** [comment]
We can have base class, but as I understand, we need to have the class in the same assembly as the db context. And also I'm not sure that will work when we use ExtCore as NuGet package or dll, not as project. So we have to test all the cases first. Could you copy your project here before creating pull request? So we could test it together. I will try to find solution too.

**Comment by Xarkam** [comment]
I've a project with @postb99 but she not push the code in github.
I will ask him to do so. It would be possible to use this base for testing.

https://github.com/OSAMES/Base

**Comment by postb99** [comment]
Hi @DmitrySikorsky and @DavidNorena , our project is up-to-date on github but on its branch. The Security extension is divided in several dlls, and then the entities are in Security.Data.Entities. Dmitry won't be surprised by the code since it's inspired by Platformus. I'm working these days a bit more and really enjoy working on it. The branch url is: https://github.com/OSAMES/Base/tree/feature/security_extension

For now the application startup performs inserts into an existing sqlite database. I wrote the entities and initialized an empty schema, all manually.

**Comment by DavidNorena** [comment]
@DmitrySikorsky sure I'll upload the code here before a PR!

**Comment by DmitrySikorsky** [comment]
So, good news is that migrations works for me now. I will try to make it more in normal way and upload here.

**Comment by DavidNorena** [comment]
I'm working on it, but didnt have success, it was failing for referenced nuget extensions .... can you share your approach ?

**Comment by DmitrySikorsky** [comment]
I think we can't use it with the migrations inside the ExtCore package, we have to have it in the real project, not package, because migrations will be created as classes in the sources. So I set the MigrationsAssembly in the UseSqlite extension method to WebApplication, and it just works. So now to make migrations work in our ExtCore-based web application we will need to
1. Add class DesignTimeStorageContextFactory (we will have parent abstract class DesignTimeStorageContextFactoryBase for that,  so it will be easy for the user).
2. We can make that ExtCore itself will find the instance of that class and initialize its static property if found one, so user will not have to do that.
3. We can make ExtCore to set the executing assembly as Migrations project automatically too. Will look at it.

So the only thing user will have to do is creating empty class in the main web application. It will be found by ExtCore, its static property will be initialized and its assembly will be used as migrations one. Trying to make it work now.

**Comment by DavidNorena** [comment]
Do we wait until you do that implementation or do you need help ?

**Comment by DavidNorena** [comment]
> We can make ExtCore to set the executing assembly as Migrations project automatically too. Will look at it.

You need to create and expose a MigrationsAssembly property is I did in my first approach PR #90 right ?

**Comment by DmitrySikorsky** [comment]
Yes, I need help. Here is the current sources. I will have to add comments,  refactor it later a bit, but now just let's test.

1. It creates migration and it's good.
2. I don't like that user will have to initialize static StorageContext property, but I don't know how to avoid it, because is is only possible to create instance of that class after options (connection string) is configured.
3. If I call Remove-Migration after it is created I have error that more than one db context found.
[ExtCore-Migrations.zip](https://github.com/ExtCore/ExtCore/files/1353028/ExtCore-Migrations.zip)

**Comment by DavidNorena** [comment]
Our problem its [here](https://github.com/aspnet/EntityFrameworkCore/blob/f4c9f4c4be0e6f159506e583eda0ba68536356b2/src/EFCore.Design/Design/Internal/DbContextOperations.cs#L128), it searches the IDesignTimeDbContextFactory implementations in the startup assembly, and there is no way to configure that behaviour, for scanning ExtCore.Data.EntityFramework.

DesignTimeStorageContextFactoryBase I think its the way to go for now.

> I don't like that user will have to initialize static StorageContext property, but I don't know how to avoid it, because is is only possible to create instance of that class after options (connection string) is configured.

Im working on that ....

**Comment by DmitrySikorsky** [comment]
I'm using Remove-Migration command from Package Manager Console. And it gives me error.

So, I  think that migrations assembly should be the main web application, do you agree with that? It can't be ExtCore.Data.EntityFramework, because in the real project we don't have sources of that project, we use it from NuGet, so there is no place to create migration classes. They should be created inside the main web app.

We can move the DesignTimeStorageContextFactory to the ExtCore.Data.EntityFramework, but in this case we (user) will have to specify which project to use to look for the db context.

So it seems that the easiest way is just to have migrations assembly set to the main web app project, and to have default startup project (again, main web application assembly), so user will have to add 1 simple class with no methods and 1 line of code (we will write static method that will do all the work). I think this part can be moved inside the AddExtCore extension method, so the only thing to do will be adding simple class. I think it is good solution?

But I still have that problem with duplicating db context on Remove-Migration.

**Comment by DavidNorena** [comment]
Ok so then ....

1. We need the user to be able to specify the name of the migrations assembly because wont be always the WebApplication Host, the user also can have a separate assembly only for models and migrations, any way give a option to specify the migrations assembly we can take my approach:

```
servicesCollection.Configure<StorageContextOptions>(options =>
            {
                options.ConnectionString = connectionString;
                options.MigrationsAssembly = migrationsAssembly;
            });
```

2. DesignTimeStorageContextFactoryBase will be on ExtCore.Data.EntityFramework project, and need to be extended by the user only for entity framework migrations be able to find the IDesignTimeContextFactory implementation.

3. Have an IServicesCollection Extension called Maybe EntityFrameworkMigrations() that will have your line of code which resolves the IStorage from services right ?.

4. dotnet ef migrations remove its not giving me errors, so we need to do the first 3 steps and then test it on package manager and dotnet tools.

Please tell me if i missed something or what needs to be changed, I think today we can have the 3.1.0 version of ExtCore with migrations support and I will help you with the docs to tell the user the only thing he has to do its call EntityFrameworkMigrations() method if he/she needs EF MigrationsSupport and also create and extend a class from DesignTimeStorageContextFactoryBase with no code, that's all !

**Comment by DavidNorena** [comment]
I'm stucked on my project because there's a lot of models and stuff, and I'm to lazy to handle Manual Sql Queries for migrations lol, but still I love your ideas and your project that's why I choosed it to be the base of my project and future projects !

**Comment by DmitrySikorsky** [comment]
So, I have prepared everything for testing, added comments, refactored the code. Now, to use migrations you only need to inherit from our abstract base class (your class will be empty) and call its method inside the Startup.ConfigureServices method. I think it is not so big deal and we can use it in this way. If you want to have separate project for the migrations, just create that class in that different project, no problems. Migrations assembly is set automatically to the project where the factory class is located.

Please, help me to fix this error. When the first migration is created, all next command will fail (both add or remove migration) with the "more than one db context found". I have that error when the Migrations folder exists. I don't know how it found it, but if I have the migrations folder and remove our class and Initialize method, I start getting error that it can't initialize StorageContext class. So it finds it somehow,  but if I remove migrations folder it doesn't see it again and says, that there is no db context found.  Very strange.

@DavidNorena, I have added your name to the copyright of the DesignTimeStorageContextFactoryBase class if you don't mind.
[ExtCore-Migrations.zip](https://github.com/ExtCore/ExtCore/files/1356102/ExtCore-Migrations.zip)

**Comment by DavidNorena** [comment]
@wow thanks for the Copyright ! 🕺 Ok man, I'll help you test and figure out why its failing

**Comment by DavidNorena** [comment]
I've got it working !

**Comment by DmitrySikorsky** [comment]
Could you upload the project? :)

**Comment by DavidNorena** [comment]
Well here is what I found:

1. In your WebApplication project you had referenced **Microsoft.EntityFrameworkCore.Tools.DotNet** and thats the CLI for **dotnet** so you can invoke it like this **dotnet ef command**, when you use Package Manager Console then you have to reference **Microsoft.EntityFrameworkCore.Tools**, so I left references to both packages so we can test both CLI's.

2. Checkout my code from the first comment in this issue and my PR, we need to add an Implementation of IDesignTimeDbContextFactory with the concrete StorageContext and not the abstract class StorageContextBase, because EF Migration tools will transverse all classes referenced to StorageContextBase thats why we had TWO context found ! (this command helped me dotnet ef dbcontext list -v).

So as i told you in my first commend we need a generic version of all implementation, that's why I changed DesignTimeStorageContextFactoryBase to be generic, the bad side is the user now have to extend that class as we talked before but also has to reference the StorageContext from the specific Provider

```
using ExtCore.Data.EntityFramework;
using ExtCore.Data.EntityFramework.Sqlite; // this line now !

namespace WebApplication
{
  public class DesignTimeStorageContextFactory : DesignTimeStorageContextFactoryBase<StorageContext>
  {
  }
}

```

3. I had a problem with this method in StorageContextBase.cs, because its hard from here infer which Provider the user its using because now the DesignTime its generic, and the method is only usefull for getting the assembly container of the DesignTimeStorageContextFactoryBase implementation, also I think StorageContextBase.cs needs to be decoupled from DesignTimeStorageContextFactoryBase as EntityFramework do.

```
protected Assembly GetMigrationsAssembly()
{
      DesignTimeStorageContextFactoryBase designTimeStorageContextFactory = ExtensionManager.GetInstance<DesignTimeStorageContextFactoryBase>();

      if (designTimeStorageContextFactory == null)
        return null;

      return designTimeStorageContextFactory.GetType().Assembly;
 }
```

I ended removing that code due to all limitations we have with this approach, and returned to my first implementation when the user now has to add one more line when configuring extcore like this:

```
  services.Configure<StorageContextOptions>(options =>
	{
	  options.ConnectionString = this.configurationRoot.GetConnectionString("Default");
	  options.MigrationsAssembly = typeof(DesignTimeStorageContextFactory).GetTypeInfo().Assembly.GetName().Name;
	}
  );
```

What do you think about it ?

Please test and give me feedback !
[ExtCore-Migrations.zip](https://github.com/ExtCore/ExtCore/files/1356757/ExtCore-Migrations.zip)

**Comment by DavidNorena** [comment]
![image](https://user-images.githubusercontent.com/26279057/31189804-e6029ff0-a8fe-11e7-9046-32dd97035abc.png)

both CLI's are working for me !

**Comment by DavidNorena** [comment]
Also please test with nuget referenced extensions

**Comment by DmitrySikorsky** [comment]
1. Microsoft.AspNetCore.All already contains, so you don't need to include it implicitly.
2. But why it works without migrations folder? First run of Add-Migration works, but it fails when there is already that folder exists. I don't really like that we need to have generic class and to specify implicitly the storage type (to have reference on the package etc).
3. Maybe you are right about decoupling. I will think.

**Comment by DmitrySikorsky** [comment]
It's very strange for me that without migrations folder it can't see any db context. But if it exists it finds one, even without any IDesignTimeDbContextFactory implementation. Why?

**Comment by DavidNorena** [comment]
1. Lol sorry I forgot that ... I don't use package manger console ...

2. I have to test and investigate that !, please take this scenario in mind ... It's not fully supported by extcore but the designtimestorage.... Generic will work ....


I have projects with multiple dbcontexts with multiple providers ... And resolving with multiple designtime... Implementations ... How so you accomplish that without generics ?

3. Yah, storage context doesn't need to know about design implementation

**Comment by DavidNorena** [comment]
It doesnt fail for me adding with folder and existing

![image](https://user-images.githubusercontent.com/26279057/31194001-5d6135be-a90b-11e7-908b-ac6d8dd1e6a6.png)

without folder ....

![image](https://user-images.githubusercontent.com/26279057/31194026-779cd51e-a90b-11e7-8366-bd32000b1c49.png)

**Comment by DmitrySikorsky** [comment]
Strange. Try to add migrations twice one by one. With different names. And remove them.

**Comment by DavidNorena** [comment]
with the last project I sent you ?

**Comment by DmitrySikorsky** [comment]
Hi! Sorry for the delay, I had too much work these days. So, I've tested your solution today, I added an initial migration, than changed the entities, added another one migration, then changed entities again and added the third one migration. Than I used Update-Database command and it worked too. Remove-Migration works as well. I changed comments a bit, but this is minor changes.

Don't forget to replace the extension DLL files in the extensions folder each time you change the entities.

I will prepare version 3.1.0-beta1 and upload it to NuGet, so we will test everything.

Good job! 👍

**Comment by DmitrySikorsky** [comment]
Also, I will write a small tutorial to the docs, so people could see how to use migrations with the ExtCore framework.

**Comment by DavidNorena** [comment]
Awesome work man !, I'll test today as soon as we know its working all, and then close this issue !

**Comment by DmitrySikorsky** [comment]
Hi! I have released ExtCore 3.1.0-beta1 with the migrations support. Also I have written small article in the docs: http://docs.extcore.net/en/latest/getting_started/using_migrations.html

Please help me to test everything. Feel free to reopen this issue (it was closed automatically) if we have any problems with this task.
